Build error against PostgreSQL 12

Hi,

pg_auto_failover cannot be built against PostgreSQL 12:

```
+ PG_CONFIG=/usr/pgsql-12/bin/pg_config
+ /usr/bin/make -j3
ipaddr.c: In function 'findHostnameFromLocalIpAddress':
ipaddr.c:548:12: warning: unused variable 'address_len' [-Wunused-variable]
  socklen_t address_len;
            ^~~~~~~~~~~
ipaddr.c:547:26: warning: unused variable 'address' [-Wunused-variable]
  struct sockaddr_storage address;
                          ^~~~~~~
health_check_worker.c: In function 'HealthCheckWorkerLauncherMain':
health_check_worker.c:233:20: error: implicit declaration of function 'AllocSetContextCreateExtended'; did you mean 'AllocSetContextCreateInternal'? [-Werror=implicit-function-declaration]
  launcherContext = AllocSetContextCreateExtended(CurrentMemoryContext,
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                    AllocSetContextCreateInternal
health_check_worker.c:233:18: error: assignment to 'MemoryContext' {aka 'struct MemoryContextData *'} from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  launcherContext = AllocSetContextCreateExtended(CurrentMemoryContext,
                  ^
health_check_worker.c: In function 'BuildDatabaseList':
health_check_worker.c:363:9: error: implicit declaration of function 'heap_beginscan_catalog'; did you mean 'table_beginscan_catalog'? [-Werror=implicit-function-declaration]
  scan = heap_beginscan_catalog(pgDatabaseRelation, 0, NULL);
         ^~~~~~~~~~~~~~~~~~~~~~
         table_beginscan_catalog
health_check_worker.c:363:7: error: assignment to 'HeapScanDesc' {aka 'struct HeapScanDescData *'} from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  scan = heap_beginscan_catalog(pgDatabaseRelation, 0, NULL);
       ^
In file included from /usr/pgsql-12/include/server/postgres.h:46,
                 from health_check_worker.c:13:
health_check_worker.c:365:49: error: passing argument 1 of 'heap_getnext' from incompatible pointer type [-Werror=incompatible-pointer-types]
  while (HeapTupleIsValid(dbTuple = heap_getnext(scan, ForwardScanDirection)))
                                                 ^~~~
/usr/pgsql-12/include/server/c.h:626:48: note: in definition of macro 'PointerIsValid'
 #define PointerIsValid(pointer) ((const void*)(pointer) != NULL)
                                                ^~~~~~~
health_check_worker.c:365:9: note: in expansion of macro 'HeapTupleIsValid'
  while (HeapTupleIsValid(dbTuple = heap_getnext(scan, ForwardScanDirection)))
         ^~~~~~~~~~~~~~~~
In file included from health_check_worker.c:21:
/usr/pgsql-12/include/server/access/heapam.h:120:45: note: expected 'TableScanDesc' {aka 'struct TableScanDescData *'} but argument is of type 'HeapScanDesc' {aka 'struct HeapScanDescData *'}
 extern HeapTuple heap_getnext(TableScanDesc scan, ScanDirection direction);
                               ~~~~~~~~~~~~~~^~~~
cli_config.c: In function 'cli_config_check':
cli_config.c:82:6: warning: unused variable 'errors' [-Wunused-variable]
  int errors = 0;
      ^~~~~~
health_check_worker.c:378:19: error: implicit declaration of function 'HeapTupleGetOid'; did you mean 'HeapTupleIsValid'? [-Werror=implicit-function-declaration]
    entry->dboid = HeapTupleGetOid(dbTuple);
                   ^~~~~~~~~~~~~~~
                   HeapTupleIsValid
health_check_worker.c:387:15: error: passing argument 1 of 'heap_endscan' from incompatible pointer type [-Werror=incompatible-pointer-types]
  heap_endscan(scan);
               ^~~~
In file included from health_check_worker.c:21:
/usr/pgsql-12/include/server/access/heapam.h:119:40: note: expected 'TableScanDesc' {aka 'struct TableScanDescData *'} but argument is of type 'HeapScanDesc' {aka 'struct HeapScanDescData *'}
 extern void heap_endscan(TableScanDesc scan);
                          ~~~~~~~~~~~~~~^~~~
health_check_worker.c: In function 'HealthCheckWorkerMain':
health_check_worker.c:452:21: error: assignment to 'MemoryContext' {aka 'struct MemoryContextData *'} from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  healthCheckContext = AllocSetContextCreateExtended(CurrentMemoryContext,
                     ^
cli_service.c: In function 'cli_monitor_run':
cli_service.c:161:16: warning: unused variable 'existingPgSetup' [-Wunused-variable]
  PostgresSetup existingPgSetup = { 0 };
                ^~~~~~~~~~~~~~~
cli_service.c:155:7: warning: unused variable 'configFilePath' [-Wunused-variable]
  char configFilePath[MAXPGPATH];
       ^~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [<builtin>: health_check_worker.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:16: monitor] Error 2
make[1]: *** Waiting for unfinished jobs....
cli_show.c: In function 'cli_show_uri':
cli_show.c:309:10: warning: unused variable 'monitor' [-Wunused-variable]
  Monitor monitor = { 0 };
          ^~~~~~~
cli_create_drop_node.c: In function 'cli_create_monitor':
cli_create_drop_node.c:427:7: warning: unused variable 'ipAddr' [-Wunused-variable]
  char ipAddr[BUFSIZE];
       ^~~~~~
cli_create_drop_node.c:425:16: warning: unused variable 'pgSetup' [-Wunused-variable]
  PostgresSetup pgSetup = config.pgSetup;
                ^~~~~~~
cli_root.c: In function 'root_options':
cli_root.c:114:6: warning: unused variable 'debugCount' [-Wunused-variable]
  int debugCount = 0;
      ^~~~~~~~~~
pgsetup.c: In function 'get_pgpid':
pgsetup.c:403:7: warning: unused variable 'line' [-Wunused-variable]
  char line[BUFSIZE];
       ^~~~
pgsetup.c:402:6: warning: unused variable 'lineno' [-Wunused-variable]
  int lineno;
      ^~~~~~
cli_do_root.c: In function 'keeper_cli_keeper_setup_getopts':
cli_do_root.c:231:23: warning: unused variable 'errors' [-Wunused-variable]
  int c, option_index, errors = 0;
                       ^~~~~~
cli_do_root.c:231:9: warning: unused variable 'option_index' [-Wunused-variable]
  int c, option_index, errors = 0;
         ^~~~~~~~~~~~
cli_do_root.c:231:6: warning: unused variable 'c' [-Wunused-variable]
  int c, option_index, errors = 0;
      ^
monitor_config.c: In function 'monitor_config_set_setting':
monitor_config.c:442:13: warning: unused variable 'option' [-Wunused-variable]
  IniOption *option;
             ^~~~~~
monitor_pg_init.c: In function 'monitor_pg_init':
monitor_pg_init.c:60:7: warning: unused variable 'postgresInstanceExists' [-Wunused-variable]
  bool postgresInstanceExists = pg_setup_pgdata_exists(&pgSetup);
       ^~~~~~~~~~~~~~~~~~~~~~
```

Can you please take a look?

Thanks, Devrim

--- a/src/bin/pg_autoctl/cli_config.c
+++ b/src/bin/pg_autoctl/cli_config.c
@@ -79,7 +79,6 @@ CommandLine config_commands =
 static void
 cli_config_check(int argc, char **argv)
 {
-	int errors = 0;
 	const bool missingPgdataIsOk = true;
 	const bool pgIsNotRunningIsOk = true;
 

--- a/src/bin/pg_autoctl/cli_create_drop_node.c
+++ b/src/bin/pg_autoctl/cli_create_drop_node.c
@@ -398,9 +398,7 @@ cli_create_monitor(int argc, char **argv)
 {
 	Monitor monitor = { 0 };
 	MonitorConfig config = monitorOptions;
-	PostgresSetup pgSetup = config.pgSetup;
 	char connInfo[MAXCONNINFO];
-	char ipAddr[BUFSIZE];
 	bool missingPgdataIsOk = true;
 	bool pgIsNotRunningIsOk = true;
 

--- a/src/bin/pg_autoctl/cli_do_root.c
+++ b/src/bin/pg_autoctl/cli_do_root.c
@@ -228,7 +228,6 @@ int
 keeper_cli_keeper_setup_getopts(int argc, char **argv)
 {
 	KeeperConfig options = { 0 };
-	int c, option_index, errors = 0;
 
 	static struct option long_options[] = {
 		{ "pgctl", required_argument, NULL, 'C' },

--- a/src/bin/pg_autoctl/cli_root.c
+++ b/src/bin/pg_autoctl/cli_root.c
@@ -112,7 +112,6 @@ CommandLine root =
 int
 root_options(int argc, char **argv)
 {
-	int debugCount = 0;
 	int verboseCount = 0;
 
 	static struct option long_options[] = {

--- a/src/bin/pg_autoctl/cli_service.c
+++ b/src/bin/pg_autoctl/cli_service.c
@@ -158,19 +158,14 @@ cli_keeper_run(int argc, char **argv)
 static void
 cli_monitor_run(int argc, char **argv)
 {
-	char configFilePath[MAXPGPATH];
 	KeeperConfig kconfig = keeperOptions;
 	MonitorConfig mconfig = { 0 };
-
 	Monitor monitor = { 0 };
 	MonitorExtensionVersion version = { 0 };
 	char postgresUri[MAXCONNINFO];
-
-	PostgresSetup existingPgSetup = { 0 };
 	bool missingPgdataIsOk = false;
 	bool pgIsNotRunningIsOk = true;
 	char connInfo[MAXCONNINFO];
-
 	char *channels[] = { "log", "state", NULL };
 
 	if (!monitor_config_init_from_pgsetup(&monitor,

--- a/src/bin/pg_autoctl/cli_show.c
+++ b/src/bin/pg_autoctl/cli_show.c
@@ -306,8 +306,6 @@ static void
 cli_show_uri(int argc, char **argv)
 {
 	KeeperConfig config = keeperOptions;
-	Monitor monitor = { 0 };
-
 	if (!IS_EMPTY_STRING_BUFFER(config.formation))
 	{
 		(void) cli_show_formation_uri(argc, argv);

--- a/src/bin/pg_autoctl/cli_systemd.c
+++ b/src/bin/pg_autoctl/cli_systemd.c
@@ -48,7 +48,7 @@ cli_systemd_getopt(int argc, char **argv)
 {
 	SystemdServiceConfig options = { 0 };
 
-	int c = 0, option_index = 0, errors = 0;
+	int c = 0, option_index = 0;
 
 	static struct option long_options[] = {
 		{ "pgdata", required_argument, NULL, 'D' },

--- a/src/bin/pg_autoctl/ipaddr.c
+++ b/src/bin/pg_autoctl/ipaddr.c
@@ -584,8 +584,6 @@ findHostnameFromLocalIpAddress(char *localIpAddress, char *hostname, int size)
 	int ret = 0;
 	char hbuf[NI_MAXHOST];
 	struct addrinfo *lookup, *ai;
-	struct sockaddr_storage address;
-	socklen_t address_len;
 
 	/* parse ipv4 or ipv6 address using getaddrinfo() */
 	ret = getaddrinfo(localIpAddress, NULL, 0, &lookup);

--- a/src/bin/pg_autoctl/monitor_config.c
+++ b/src/bin/pg_autoctl/monitor_config.c
@@ -445,7 +445,6 @@ monitor_config_set_setting(MonitorConfig *config,
 						   char *value)
 {
 	const char *filename = config->pathnames.config;
-	IniOption *option;
 	IniOption monitorOptions[] = SET_INI_OPTIONS_ARRAY(config);
 
 	if (ini_set_setting(filename, monitorOptions, path, value))

--- a/src/bin/pg_autoctl/monitor_pg_init.c
+++ b/src/bin/pg_autoctl/monitor_pg_init.c
@@ -57,7 +57,6 @@ monitor_pg_init(Monitor *monitor, MonitorConfig *config)
 	char configFilePath[MAXPGPATH];
 	char postgresUri[MAXCONNINFO];
 	PostgresSetup pgSetup = config->pgSetup;
-	bool postgresInstanceExists = pg_setup_pgdata_exists(&pgSetup);
 
 	if (directory_exists(pgSetup.pgdata))
 	{

--- a/src/bin/pg_autoctl/pgsetup.c
+++ b/src/bin/pg_autoctl/pgsetup.c
@@ -410,8 +410,6 @@ static bool
 get_pgpid(PostgresSetup *pgSetup, bool pg_is_not_running_is_ok)
 {
 	FILE *fp;
-	int lineno;
-	char line[BUFSIZE];
 	char pidfile[MAXPGPATH];
 	long pid = -1;
 

--- a/src/monitor/health_check_worker.c
+++ b/src/monitor/health_check_worker.c
@@ -230,11 +230,11 @@ HealthCheckWorkerLauncherMain(Datum arg)
 	/* Make background worker recognisable in pg_stat_activity */
 	pgstat_report_appname("pg_auto_failover monitor launcher");
 
-	launcherContext = AllocSetContextCreateExtended(CurrentMemoryContext,
-													"Health Check Launcher Context",
-													ALLOCSET_DEFAULT_MINSIZE,
-													ALLOCSET_DEFAULT_INITSIZE,
-													ALLOCSET_DEFAULT_MAXSIZE);
+	launcherContext = AllocSetContextCreate(CurrentMemoryContext,
+											"Health Check Launcher Context",
+											ALLOCSET_DEFAULT_MINSIZE,
+											ALLOCSET_DEFAULT_INITSIZE,
+											ALLOCSET_DEFAULT_MAXSIZE);
 
 	MemoryContextSwitchTo(launcherContext);
 
@@ -352,7 +352,7 @@ BuildDatabaseList(void)
 {
 	List	   *databaseList = NIL;
 	Relation	pgDatabaseRelation;
-	HeapScanDesc scan;
+	TableScanDesc scan;
 	HeapTuple	dbTuple;
 	MemoryContext originalContext = CurrentMemoryContext;
 
@@ -360,7 +360,7 @@ BuildDatabaseList(void)
 
 	pgDatabaseRelation = heap_open(DatabaseRelationId, AccessShareLock);
 
-	scan = heap_beginscan_catalog(pgDatabaseRelation, 0, NULL);
+	scan = table_beginscan_catalog(pgDatabaseRelation, 0, NULL);
 
 	while (HeapTupleIsValid(dbTuple = heap_getnext(scan, ForwardScanDirection)))
 	{
@@ -449,11 +449,11 @@ HealthCheckWorkerMain(Datum arg)
 	 * Only process given database when the extension has been loaded.
 	 * Otherwise, happily quit.
 	 */
-	healthCheckContext = AllocSetContextCreateExtended(CurrentMemoryContext,
-													   "Health check context",
-													   ALLOCSET_DEFAULT_MINSIZE,
-													   ALLOCSET_DEFAULT_INITSIZE,
-													   ALLOCSET_DEFAULT_MAXSIZE);
+	healthCheckContext = AllocSetContextCreate(CurrentMemoryContext,
+											   "Health check context",
+											   ALLOCSET_DEFAULT_MINSIZE,
+											   ALLOCSET_DEFAULT_INITSIZE,
+											   ALLOCSET_DEFAULT_MAXSIZE);
 
 	MemoryContextSwitchTo(healthCheckContext);
 

--- a/src/monitor/replication_state.c
+++ b/src/monitor/replication_state.c
@@ -16,6 +16,7 @@
 
 #include "metadata.h"
 #include "replication_state.h"
+#include "version_compat.h"
 
 #include "access/htup.h"
 #include "access/htup_details.h"

--- a/src/monitor/version_compat.h
+++ b/src/monitor/version_compat.h
@@ -27,12 +27,24 @@
 #define BackgroundWorkerInitializeConnectionByOid(dboid, useroid, flags) \
 	BackgroundWorkerInitializeConnectionByOid(dboid, useroid)
 
-static inline MemoryContext
-AllocSetContextCreateExtended(MemoryContext parent, const char *name, Size minContextSize,
-							  Size initBlockSize, Size maxBlockSize)
+#endif
+
+#if (PG_VERSION_NUM < 120000)
+
+#define table_beginscan_catalog heap_beginscan_catalog
+#define TableScanDesc HeapScanDesc
+
+#endif
+
+#if (PG_VERSION_NUM >= 120000)
+
+#include "access/htup_details.h"
+#include "catalog/pg_database.h"
+
+static inline Oid HeapTupleGetOid(HeapTuple tuple)
 {
-	return AllocSetContextCreate(parent, name, minContextSize, initBlockSize,
-								 maxBlockSize);
+	Form_pg_database dbForm = (Form_pg_database) GETSTRUCT(tuple);
+	return dbForm->oid;
 }
 
 #endif

--- a/src/monitor/version_compat.h
+++ b/src/monitor/version_compat.h
@@ -14,7 +14,12 @@
 
 #include "postgres.h"
 
-#if (PG_VERSION_NUM >= 90600 && PG_VERSION_NUM < 110000)
+/* we only support PG10 and PG11, PG 12 is not supported yet*/
+#if (PG_VERSION_NUM < 100000 || PG_VERSION_NUM >= 120000)
+#error "Unknown or unsupported postgresql version"
+#endif
+
+#if (PG_VERSION_NUM < 110000)
 
 #include "postmaster/bgworker.h"
 #include "utils/memutils.h"

